### ISSUE TITLE
How to speed up rendering of many plots

### ISSUE THREAD

**Issue by EmanuelFeru** [ScottPlot/ScottPlot#1127]
**ScottPlot Version:** 4.1.16
**Operating System:** WIN10
**Application Type:** WinForms

**Issue Description:** In my application I create a list of plots that are linked. Basically I allow the user to create as many plots as he wants. The drawback is that with the increased number of plots and signals the interaction with the plots becomes slower and slower.

**Questions:** Do you have suggestions on what to do to speed up the rendering process? (compute in parallel, threads, use GPU, etc.)

Below is a snippet of the code that is excuted every time the use interracts with one of the plots. Works decent up to 4 plots and moderate number of signals (up to 10 signals/plot, and let's say up to 10000 points/signal). However, above that I start to feel degradation of speed that affects user interraction. 
Using `Parallel.ForEach` doesn't help much. Especially for the `Vertical Lines dragged linked` I get strange delayed behavior if I use  `Parallel.ForEach`.
* In the past I have used a graph tool that used the [Arction engine](https://www.arction.com/lightningchart-ultimate-sdk/). It was very fast, I must admit even, with many many plots. It is using DirectX in the background. Maybe we can learn from it.
* Then I was triggered by Matlab, which I intensively use. Matlab uses OpenGL to render and it renders quite fast. Do you think OpenGL would increase rendering time in C#, see:
https://github.com/dwmkerr/sharpgl
If we could use it in ScottPlot, then 3D plots and surface plots will become possible too. What is your opinion on that?
This topic is related to #1113 

* Or is there a posibility to use the GPU power in C# for rendering?
* For me any option that helps to speed up the rendering of many plots is ok. Because only 20% of processing power is used at the moment on my laptop. So, I think we can do more to use the available resources to our advantage.

Other links that I found:
* https://github.com/sharpdx/SharpDX
* https://github.com/microsoft/Win2D
* https://www.reddit.com/r/csharp/comments/5uvyw9/what_can_i_use_for_basic_fast_2d_graphics/?utm_source=amp&utm_medium=&utm_content=post_body

**For X Axes linked:**
```cs
public static void OnAxesChanged(object sender, EventArgs e)
{
    FormsPlot changedPlot = (FormsPlot)sender;
    axisXMin = changedPlot.Plot.GetAxisLimits().XMin;
    axisXMax = changedPlot.Plot.GetAxisLimits().XMax;

    foreach (FormTimePlot tp in timePlot)
    {
        if (tp.formsPlot == changedPlot)
            continue;

        // Disable events briefly to avoid an infinite loop
        tp.formsPlot.Configuration.AxesChangedEventEnabled = false;
        tp.formsPlot.Plot.SetAxisLimits(xMin: axisXMin, xMax: axisXMax);
        tp.formsPlot.Render();
        tp.formsPlot.Configuration.AxesChangedEventEnabled = true;
    }

    //Parallel.ForEach (timePlot, tp =>
    //{
    //    // Disable events briefly to avoid an infinite loop
    //    tp.formsPlot.Configuration.AxesChangedEventEnabled = false;
    //    tp.formsPlot.Plot.SetAxisLimits(xMin: axisXMin, xMax: axisXMax);
    //    tp.formsPlot.Render();
    //    tp.formsPlot.Configuration.AxesChangedEventEnabled = true;
    //});

}
```

**For Vertical Lines dragged linked:**
```cs
public static void OnAxesDragged(object sender, EventArgs e)
{
    VLine changed_Line = (VLine)sender;
    vPosition = changed_Line.X;

    foreach (FormTimePlot tp in timePlot)
    {
        if (tp.vLine == changed_Line)
            continue;

        tp.vLine.X = vPosition;
        tp.formsPlot.Configuration.AxesChangedEventEnabled = false;
        tp.formsPlot.RenderRequest();
        tp.formsPlot.Configuration.AxesChangedEventEnabled = true;                
    }

    //Parallel.ForEach(timePlot, tp =>
    //{
    //    tp.vLine.X = vPosition;
    //    tp.formsPlot.Configuration.AxesChangedEventEnabled = false;
    //    tp.formsPlot.Render();
    //    tp.formsPlot.Configuration.AxesChangedEventEnabled = true;
    //});
}
```

**Comment by bclehmann** [comment]
One thing that might help instead of `Parallel.ForEach` is several `Task.Run`s. Or writing your own multithreaded code. 

`Parallel.ForEach` has weird characteristics, it generally uses as many threads as reasonably possible, and those threads are often not that busy. This means that if your CPU has other things to do they may not be scheduled as often. It also may spawn more threads than is efficient for a given workload. You can also pass in a `ParallelOptions` and play with `MaxDegreeOfParallelism` (or `TaskScheduler` if you want to get that low-level) and see if that helps.

As for ScottPlot, there's been talk for a while of moving to Skia, which is anticipated to have better performance than System.Drawing. I also wonder if ScottPlot may perform differently on Linux or Mac as libgdiplus uses Cairo which uses OpenGL (or sometimes another backend), and I don't know what GDI+ uses on Windows. There was a hacky way of forcing a program to use a dedicated GPU discussed in #413, however, that only works on Windows systems with NVIDIA GPUs (although on other systems it shouldn't break anything). Note that ScottPlot will still use hardware acceleration regardless, just not necessarily on the primary GPU.

There are general optimizations that could be done within ScottPlot. I don't think that there are many algorithms that are majorly unperformant, at least from a time complexity perspective. However, I don't touch signal plotting very often, and that would be the most performance-sensitive area. But in C# things like reducing heap allocations can be big performance wins, even if an algorithm is theoretically optimal. That's why things like LINQ are controversial among some developers.

Another optimization was discussed in #978. Unfortunately, it's not really feasible with GDI+ as GDI objects must only be used on the thread they're created on. It may be worth considering with Skia, although Skia may use object pools internally.

**Comment by StendProg** [comment]
Hi @EmanuelFeru, @bclehmann,

A lot of performance optimizations have been made.
In particular, Signals use multithreading by default, which is why the additional Parallel loop does little to nothing.
And at the moment, in fact, the bottleneck in ScottPlot is GDI+. It does not use hardware acceleration to draw lines in any way. Therefore, each signal from its 1000+ lines adds its additional 10-50 milliseconds to rendering.

The transition to another library is always in my head, but all the time it is postponed. Since this is not a simple matter, and the current performance of the ScottPlot is quite satisfactory for most even heavily loaded applications.

>As for ScottPlot, there's been talk for a while of moving to Skia, which is anticipated to have better performance than System.Drawing.

There is a positive experience with `SkiaSharp` #114, but it is already very outdated, at the moment there may be something simpler and more interesting.

**Comment by EmanuelFeru** [comment]
I will do some more tries. For now I will close the topic. Thank you for your answers.

**Comment by EmanuelFeru** [comment]
Hi Scott,
It's really good to see that you found time to work on Scottplot again. I'm reopening this topic because I just read your full discussion in #114 and in my opinion there is a lot of potential in using something like [SkiaSharp](https://github.com/mono/SkiaSharp)  or [SharpDX](https://github.com/sharpdx/SharpDX) 

As I said I used a tool in the past based on [SharpDX](https://github.com/sharpdx/SharpDX) and worked fantastically well, even with huge data points and many signal. On top, no difference was felt when going to large screens.

At the moment, with GDI+ I do feel a drop in speed when I go or maximize to a big screen, even with 0 signals in the plot (only my usual vertical Draggable line).

It would be great I would really like to encourage to continue on [SkiaSharp](https://github.com/mono/SkiaSharp), I am sure something great will come out. Processing power is there we just need to use it. I can be the first one using ScottPlotSkia.

**Comment by EmanuelFeru** [comment]
It is quite amazing what it can do:
https://skia.org/docs/user/modules/particles/

Let's resume Skia developments :)

**Comment by EmanuelFeru** [comment]
Linked to #1036

**Comment by swharden** [comment]
@EmanuelFeru a while back I did some experimentation with SkiaSharp and plotting and it was fantastic performance https://github.com/swharden/quickplot

The problem at that time was the build process was non-trivial (compared to System.Drawing which just seemed to work everywhere). 

With #1036 and discussion emerging there this issue is on its way to being solved! I'll close this issue for now, and hopefully tackle this around the time .NET 6 and `Microsoft.Maui.Graphics.Skia` is out of preview 👍
